airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py

```diff
@@ -434,9 +434,9 @@ def sync(self) -> None:
                     )
                     self.fail(task[0], e)
                 except ApiException as e:
-                    # These codes indicate something is wrong with pod definition; otherwise we assume pod
-                    # definition is ok, and that retrying may work
-                    if e.status in (400, 422):
+                    # In case of the below error codes, fail the task and honor the task retires.
+                    # Otherwise, go for continuous/infinite retries.
+                    if e.status in (400, 403, 404, 422):
```

Comment:
Personally, I prefer to handle 403 as a special case, with a user-provided number of attempts, -1 by default to trying indefinitely (the current behavior), or trying x times depending on the provided configuration, which may be zero. WDYT?

Reply:
I am not in favor of overloading the Kubernetes API server by default.  Even in case of transient errors like quota exceeds, it needs some time for quota to be available. In certain scenarios like permission issues, the task ends in a queued state forever which is not good. Instead, let the task fail eventually after retries so that the consumer can take an appropriate action (increase quota or adjust the job timing). 
         - 403 Forbidden will returns in scenarios like
             - your request exceeds the namespace quota 
             - scheduler role doesn't have permission to launch the pod

Reply:
Maybe we should wait between retries when 403 occurs.
This way, we won't overload the K8s API and also current implementation will be supported (though in a slightly different way).
However, this will probably complicate the current implementation.

Reply:
> it needs at-least a few mintues for quota to be available

Then we can also support a backoff retry strategy with the max retries number. I'm just trying to avoid failing an Airflow task because of a temporary pressure on the cluster, especially if the users use a small number of retries for the task. Increasing the number of retries to avoid failing the task when the quota is exceeded will lead to useless attempts when there is a problem in the task.

Reply:
Yes. Backoff retry strategy works for transient errors. However, it doesn't make sense to apply for non-transient errors. K8 API error code 403 includes both transient and non-transient errors. We need to deep-dive into all the error messages and apply a back-off retry strategy for transient errors only. However, it is a bit more complicated but doable.

Reply:
I think using normal task retries is okay. Folks can always use pools to spread out big workloads.

Even if we did do another retry counter for this, it still should eventually fail. And I'm just not sure this deserves a separate retry counter 🤷.

Reply:
> I think using normal task retries is okay. Folks can always use pools to spread out big workloads.
> 
> Even if we did do another retry counter for this, it still should eventually fail. And I'm just not sure this deserves a separate retry counter 🤷.

@jedcunningham Yes. I am with you on this. @hussein-awala and @shohamy7 Let us know your opinion on this.

Reply:
@jedcunningham Currently, when creating a worker pod fails due to quota exceeding, which is a temporary failure, the executor retries again and again until some resources are free.

This PR updates this to fail the task once the pod creation fails due to exceeding quota, which could be considered a breaking (or significant) change for users who have a quota set up and have a small number of attempts. In addition, it does not offer the possibility to update the new behaviour to retry creating the pod without causing the task to fail, so the only solution will be to increase the number of attempts, which will be used regardless of the type failure, and will add additional latency (retry_delay, retry_exponential_backoff, ...) to the task execution.

> Even if we did do another retry counter for this, it still should eventually fail.

The quota is the sum of resources used by all the pods in a namespace, so when other pods terminate, some resources will be free, and creating a new pod will be possible.

Reply:
Ideally, the scheduler pool slots should be mapped to the namespace quota. So, that we don't end up in quota exceeds exception. Even if they deviate over time, the customer should be aware of the failures instead of masking/hiding the failures. So, that the customer can adjust the pool slots as per the quota. I believe this problem falls under resource-aware scheduling and we have to think about it holistically across multiple executors.

Reply:
Maybe this can be configurable via `airflow.cfg` under the `kubernetes_executor` and setting the default as the current implementation?
I agree with @hussein-awala that this is a breaking (or significant) change, but I also understand the problem of relying on the `KubernetesExecutor` to infinite retry on such errors.
I am not sure what you are thinking about adding more configuration to the kubernetes executor, but maybe it may be a good settlement by letting the users to decide how they want the executor will deal with this use case.

Reply:
Yes. It's breaking change. If we do the right thing, then I believe it's okay to introduce it in the major version release. 
The question is about retaining the old behavior, not the implementation.

Reply:
@jedcunningham and @hussein-awala 
Let us know the final thoughts on this. Based on that we can take next steps.

Reply:
> when creating a worker pod fails due to quota exceeding, which is a temporary failure, the executor retries again and again until some resources are free.

I view that behavior is a bug.

> The quota is the sum of resources used by all the pods in a namespace, so when other pods terminate, some resources will be free, and creating a new pod will be possible.

Not necessarily. The pod could be too large to ever be created, and it's still stuck in the loop forever.

I still think it's better to use the existing retry mechanism for this, I think this is the "right" behavior even if it's significantly different that what it is right now.

That said, we do have precedent for the config approach with [celerys task_publish_max_retries](https://github.com/apache/airflow/blob/6629e2bf7b80d07f2cf61895873521400cdb0d5b/airflow/providers/celery/provider.yaml#L287). If we go this route though, I really hope we can be targeted to quota failures specifically, and have a default of 0.

Reply:
> Not necessarily. The pod could be too large to ever be created, and it's still stuck in the loop forever.

Yeah, it could be too large or not, but with a new config, we can give the user the ability to decide what to do, I'm fine with a new config similar to `task_publish_max_retries` even if it's 0 by default, but at least the user could increase this to a higher value.

Reply:
How does this compare to just increasing your quota in kubernetes and let the K8s scheduler handle all the queuing? Pods can be in queued state without using any node compute.
Stacking queues and schedulers on top of each other feels like just pushing the demand elsewhere.

For example, if retry is implemented in airflow, will the tasks be queued fairly into kubernetes or will new tasks get precedence? 
Eg 
t1: Task 2 attempts to start. Quota is full.  Fails -> Backoff 5 min
t2: Task 1 completes -> Quota is now available
t3: Task 3: Starts -> Succeeds because no quota limitation any more
t4: Task 2 retries, quota is full again, because Task3 squeezed in before. Not fair because it should have gone before Task 3.


To be honest i'm not very familiar with quotas, it seems like something that is more useful for servers that are running all the time, not batch-oriented workflow like airflow tasks that start, stop all the time and where relying on the built in queue may be more appropriate? Maybe others have valid use cases for it i'm not aware of.

Reply:
@jedcunningham and @hussein-awala 
I have added the support for task_publish_max_retries. Can you review it?

Reply:
@hussein-awala Can you review it?

Reply:
I have addressed the review comments. Can you review it?